src/main/scala/org/constellation/invertedmap/InvertedMap.scala

```diff
@@ -0,0 +1,54 @@
+package org.constellation.invertedmap
+
+import scala.collection.generic.{CanBuildFrom, GenMapFactory}
+import scala.collection.immutable.MapLike
+import scala.collection.{GenTraversableOnce, mutable}
+
+final class InvertedMap[A, B] private (m: Map[B, Set[A]])
```

Comment:
Or another idea worth considering `Map[Either[A, B], Either[B, Set[A]]`

Reply:
The idea is to reduce amount of data needed to store which we can use in at least 2 different places.
Consider such a use case. We have following observations `Map[Id, NodeState]` from every node, so `Map[Id, Map[Id, NodeState]]`. For every node id we have to store its observations. 99% time observations should be equal, so we store duplications unnecessary.
Instead of that, we can store it in inverted way `Map[Map[Id, NodeState], Set[Id]` so we store observations once for associated set of ids which holds same observations. To simplify interactions with such a map - there is an InvertedMap which extends scala collections and allow to use common Map methods.
That being said, you can use it like that:
```
val m = InvertedMap(("node1", allReady), ("node2", allReady), ("node3", oneOffline))
m.get("node2")
m.updated("node3", oneOffline)
m - "node3"
```
but under the hood, data are not duplicated

Reply:
Even if values are large, you're operating on hashes and pointers, so additional map shouldn't introduce too much overhead, but that depends on the number items stored.

Also since the lookups are linear, creating inverted map should be the same as exposing an inverted map with extension method I think `def inverted(): Map[B, Set[A]] = m.groupBy(_._2).view.mapValues(_.keys)`

Reply:
but you would have to map it back and forth. it does the same but keeps the underlying map under the hood

Reply:
That's true. One more thing that might be considered is using object-pool for `NodeState` with underlying `WeakHashSet` (a set which values can be GC'ed if there are no other references to it).

Reply:
NodeState is just an example of use case

Reply:
Yep, I understand that, and you are doing object-pooling either way here, but it's just not visible. So I still think that the explicit object-pool would be a better choice (if the goal is to reduce the memory usage).